File: app/src/main/java/com/scop/org/minesweeper/generators/solver/BasicSolver.java
package com.scop.org.minesweeper.generators.solver;

import com.scop.org.minesweeper.elements.Tile;
import com.scop.org.minesweeper.utils.GridUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

import static com.scop.org.minesweeper.elements.Tile.Status.COVERED;

public class BasicSolver extends Solver {

	@Override
	public boolean analyze(){
		boolean changesMade = false;

		List<Tile> sketchTiles = sketch.getTiles();
		List<Integer> sketchTilesUncovered = new ArrayList<>();

		for (int i : sketch.getUncovered()) {
			Tile sTile = sketchTiles.get(i);

			if (sTile.getBombsNear() > 0){
				List<Integer> nearCovered = GridUtils.getNeighborsIdx(sketch.getGrid(), sTile).stream()
						.filter(ix->sketchTiles.get(ix).getStatus() == COVERED)
						.collect(Collectors.toList());

				if (sTile.getBombsNear() == nearCovered.size()){
					changesMade |= nearCovered.size() > 0;
					nearCovered.forEach(this::markBomb);
				}
			} else {
				sketchTilesUncovered.add(i);
			}
		}

		for (int i : sketchTilesUncovered) {
			Tile sTile = sketchTiles.get(i);

			if (sTile.hasCustomFlag('0')) {

				for (int idx : GridUtils.getNeighborsIdx(sketch.getGrid(), sTile) ){
					if (sketchTiles.get(idx).getStatus() == COVERED){
						reveal(idx);
						changesMade = true;
					}
				}
			}
		}


		return changesMade;
	}
}

File: app/src/main/java/org/oar/minesweeper/generators/solver/BasicSolver.kt
package org.oar.minesweeper.generators.solver

import org.oar.minesweeper.models.TileStatus
import org.oar.minesweeper.utils.GridUtils.getNeighborsIdx

class BasicSolver : Solver() {

    override fun analyze(): Boolean {
        var changesMade = false
        val sketchTiles = sketch.tiles
        val sketchTilesUncovered: MutableList<Int> = ArrayList()
        for (i in sketch.uncovered) {
            val sTile = sketchTiles[i]
            if (sTile.bombsNear > 0) {
                val nearCovered = sketch.grid.getNeighborsIdx(sTile)
                    .filter { ix -> sketchTiles[ix].status === TileStatus.COVERED }

                if (sTile.bombsNear == nearCovered.size) {
                    changesMade = changesMade or (nearCovered.isNotEmpty())
                    nearCovered.forEach { idx -> markBomb(idx) }
                }
            } else {
                sketchTilesUncovered.add(i)
            }
        }

        for (i in sketchTilesUncovered) {
            val sTile = sketchTiles[i]
            if (sTile.customFlag.contains('0')) {
                for (idx in sketch.grid.getNeighborsIdx(sTile)) {
                    if (sketchTiles[idx].status === TileStatus.COVERED) {
                        reveal(idx)
                        changesMade = true
                    }
                }
            }
        }

        return changesMade
    }
}